Mathlib/ModelTheory/Basic.lean

```diff
@@ -0,0 +1,1100 @@
+    (#(L.sum L').Functions i)
+      = (Cardinal.lift.{u'} (#L.Functions i) + Cardinal.lift.{u} (#L'.Functions i) : Cardinal) := by
+  simp [Language.sum]
+#align first_order.language.card_functions_sum FirstOrder.Language.card_functions_sum
+
+@[simp]
+theorem card_relations_sum (i : ℕ) :
+    (#(L.sum L').Relations i) =
+      Cardinal.lift.{v'} (#L.Relations i) + Cardinal.lift.{v} (#L'.Relations i) := by
+  simp [Language.sum]
+#align first_order.language.card_relations_sum FirstOrder.Language.card_relations_sum
+
+@[simp]
+theorem card_sum :
+    (L.sum L').card = Cardinal.lift.{max u' v'} L.card + Cardinal.lift.{max u v} L'.card := by
+  simp only [card_eq_card_functions_add_card_relations, card_functions_sum, card_relations_sum,
+    sum_add_distrib', lift_add, lift_sum, lift_lift]
+  simp only [add_assoc, add_comm (Cardinal.sum fun i => (#L'.Functions i).lift)]
+#align first_order.language.card_sum FirstOrder.Language.card_sum
+
+@[simp]
+theorem card_mk₂ (c f₁ f₂ : Type u) (r₁ r₂ : Type v) :
+    (Language.mk₂ c f₁ f₂ r₁ r₂).card =
+      Cardinal.lift.{v} (#c) + Cardinal.lift.{v} (#f₁) + Cardinal.lift.{v} (#f₂) +
+          Cardinal.lift.{u} (#r₁) + Cardinal.lift.{u} (#r₂) := by
+  simp [card_eq_card_functions_add_card_relations, add_assoc]
+
+#align first_order.language.card_mk₂ FirstOrder.Language.card_mk₂
+
+variable (L) (M : Type w)
+
+/-- A first-order structure on a type `M` consists of interpretations of all the symbols in a given
+  language. Each function of arity `n` is interpreted as a function sending tuples of length `n`
+  (modeled as `(Fin n → M)`) to `M`, and a relation of arity `n` is a function from tuples of length
+  `n` to `Prop`. -/
+@[ext]
+class Structure where
+  /-- Interpretation of the function symbols -/
+  funMap : ∀ {n}, L.Functions n → (Fin n → M) → M
+  /-- Interpretation of the relation symbols -/
+  rel_map : ∀ {n}, L.Relations n → (Fin n → M) → Prop
+set_option linter.uppercaseLean3 false in
+#align first_order.language.Structure FirstOrder.Language.Structure
+
+variable (N : Type w') [L.Structure M] [L.Structure N]
+
+open Structure
+
+/-- Used for defining `FirstOrder.Language.Theory.Model.inhabited`. -/
+def inhabited.trivialStructure {α : Type _} [Inhabited α] : L.Structure α :=
+  ⟨default, default⟩
+#align
+  first_order.language.inhabited.trivial_structure
+  FirstOrder.Language.inhabited.trivialStructure
+
+/-! ### Maps -/
+
+
+/-- A homomorphism between first-order structures is a function that commutes with the
+  interpretations of functions and maps tuples in one structure where a given relation is true to
+  tuples in the second structure where that relation is still true. -/
+structure Hom where
+  /-- The underlying function of a homomorphism of structures -/
+  toFun : M → N
+  /-- The homomorphism commutes with the interpretations of the function symbols -/
+  map_fun' : ∀ {n} (f : L.Functions n) (x), toFun (funMap f x) = funMap f (toFun ∘ x) := by
+    intros; trivial
```

Comment:
We should make all the autoparams here `aesop`, I think.

See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Aesop.20and.20cases

Reply:
Please check the commit right below. Changing the corresponding line to `aesop` fails at some point. Also I'm not familiar with `aesop`.